internal/keyring/keyring.go

```diff
@@ -0,0 +1,80 @@
+// Package keyring is a simple wrapper that adds timeouts to the zalando/go-keyring package.
+package keyring
+
+import (
+	"context"
+	"time"
+
+	"github.com/zalando/go-keyring"
+)
+
+type TimeoutError struct {
+	message string
+}
+
+func (e *TimeoutError) Error() string {
+	return e.message
+}
+
+// Set secret in keyring for user.
+func Set(service, user, secret string) error {
+	duration := time.Second
+	ctx, cancel := context.WithTimeout(context.Background(), duration)
+	defer cancel()
+	ch := make(chan error, 1)
+	go func() {
+		defer close(ch)
+		ch <- keyring.Set(service, user, secret)
+	}()
+	select {
+	case err := <-ch:
+		return err
+	case <-ctx.Done():
+		return &TimeoutError{"timeout while trying to set secret in keyring"}
+	}
+}
+
+// Get secret from keyring given service and user name.
+func Get(service, user string) (string, error) {
```

Comment:
I'm trying to figure out if this function is racy or not. See this playground: https://go.dev/play/p/hCpEywlYjiR which you can run a few times to see different results).

I **think** it is racy, like the example in the playground, but I can't quite convince myself that `time.Sleep` doesn't have some impact on goroutine scheduling. Either way, unless you're sure this code isn't racy then I think we should change it. At the heart of the issue is that `select` can choose any channel branch that succeeds, and putting a value on one channel can race with closing the other, e.g. `errCh <- err` racing with `close(resCh)`.

There are a few possible (I think) fixes for this:
 1. Using unbuffered channels will block the write until someone reads, preventing the `close` until a branch has already been picked
 2. Move the `closes` until after a decision has been made
 3. Use a result struct `type result struct { secret string, error string }` or anonymous struct `c := make(chan struct { string; error })` and only have a single channel
 
I would favour 3, as any use of multiple channels has the propensity to introduce these kinds of hard to reason about issues and there's probably a reason I've seen it used so many times.

On the other hand, if you know a reason this isn't racy then please hit me with that knowledge!

Reply:
Great catch here, I do think there is a race condition! I had not thought about how `select` would work with `defer close`. I was thinking about using option 3 originally, seems like the best solution, I will make the change.